src/17_integrate_tox21_12.py

```diff
@@ -0,0 +1,93 @@
+
+# Properties table
+column_descriptions = {
+    "NR-AR": "Androgen Receptor assay tests for agonists of the androgen receptor.",
+    "NR-AR-LBD": "Androgen Receptor Ligand Binding Domain assay assesses the ability of chemicals to bind to the androgen receptor.",
+    "NR-AhR": "Aryl Hydrocarbon Receptor assay tests for activation of the aryl hydrocarbon receptor.",
+    "NR-Aromatase": "Aromatase assay tests for inhibition of aromatase, an enzyme responsible for a key step in estrogen biosynthesis.",
+    "NR-ER": "Estrogen Receptor assay tests for agonists of the estrogen receptor.",
+    "NR-ER-LBD": "Estrogen Receptor Ligand Binding Domain assay assesses the binding ability of chemicals to the estrogen receptor.",
+    "NR-PPAR-gamma": "Peroxisome Proliferator-Activated Receptor Gamma assay targets PPAR-gamma receptor, regulating fatty acid storage and glucose metabolism.",
+    "SR-ARE": "Antioxidant Response Element assay identifies chemicals that activate ARE, involved in response to oxidative stress.",
+    "SR-ATAD5": "ATAD5 assay assesses DNA damage by identifying chemicals that increase ATAD5 expression, involved in DNA repair.",
+    "SR-HSE": "Heat Shock Element assay identifies chemicals that activate HSE, involved in protecting cells from stress.",
+    "SR-MMP": "Matrix Metalloproteinase assay tests for inhibition of matrix metalloproteinases, involved in extracellular matrix breakdown.",
+    "SR-p53": "p53 assay identifies chemicals that activate p53, a protein crucial for controlling cell division and apoptosis."
+}
+
+properties_list = []
+for idx, (col_name, description) in enumerate(column_descriptions.items()):
+    data_json = json.dumps({
+        "property": col_name,
+        "active_value": 1,
+        "inactive_value": 0
+    })
+    properties_list.append((str(idx), data_json))  # Ensure pid is a string
+
+properties_df = spark.createDataFrame(properties_list, ["pid", "data"])
+properties_df.write.mode("overwrite").parquet("staging/Tox21/properties.parquet")
+
+# Activities table
+activities = []
+for idx, col_name in enumerate(column_descriptions.keys()):
+    activity_table = tox21_with_inchi.withColumn("pid", F.lit(str(idx))) \
+                                     .withColumn("value", F.col(col_name)) \
+                                     .withColumn("aid", F.monotonically_increasing_id().cast("string")) \
+                                     .withColumn("source", F.lit("Tox21")) \
+                                     .select("aid", "sid", "pid", "smiles", "inchi", "source", "value")
+    activities.append(activity_table)
+
+final_activity_table = activities[0]
+for activity_table in activities[1:]:
+    final_activity_table = final_activity_table.union(activity_table)
+
+# Ensure unique 'aid' by assigning row numbers
+window_spec = Window.orderBy(F.monotonically_increasing_id())
+final_activity_table = final_activity_table.withColumn("aid", row_number().over(window_spec).cast("string"))
```

Comment:
this can be done more concisely:

final_activity_table = final_activity_table.withColumn("aid", F.monotonically_increasing_id().cast("string"))